Make mir borrowck's use of opaque types independent of the typeck query's result

oli-obk opened the pull request:
fixes #87218
fixes #86465

we used to use the typeck results only to generate an obligation for the mir borrowck type to be equal to the typeck result.

When i removed the `fixup_opaque_types` function in #87200, I exposed a bug that showed that mir borrowck can't doesn't get enough information from typeck in order to build the correct lifetime mapping from opaque type usage to the actual concrete type. We therefor now fully compute the information within mir borrowck (we already did that, but we only used it to verify the typeck result) and stop using the typeck information.

We will likely be able to remove most opaque type information from the borrowck results in the future and just have all current callers use the mir borrowck result instead.

r? @spastorino

oli-obk committed:
Fix VecMap::iter_mut

It used to allow you to mutate the key, even though that can invalidate the map by creating duplicate keys.

oli-obk committed:
Make mir borrowck's use of opaque types independent of the typeck query's result

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3106241951) [(plain)](https://github.com/rust-lang/rust/commit/1690f7e08d35ef577887e2ecd91cc3cac8c50f64/checks/3106241951/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* 340 features
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:13
```

</details>

lqd commented:
By the way oli, does this fix the similar ICE in this slightly-modified version of #87258 (to remove the syntax error you saw but keep the ICE) ? (Seems unlikely, but it doesn't hurt to ask :)

```rust
#![feature(min_type_alias_impl_trait)]
#![feature(generic_associated_types)]
use std::future::Future;

pub trait Trait1
{
    type Error: Clone;
    type Assoc2<'a>
    where
        Self: 'a;
}

pub trait Trait2: Trait1 {
    type FooFuture<'a, 'b>: Future<Output = Result<(), Self::Error>>;
    fn foo<'a: 'b, 'b>(assoc2: &'b mut Self::Assoc2<'a>) -> Self::FooFuture<'a, 'b>
    where
        Self: 'a;
}

impl<'c, S : Trait1> Trait1 for &'c mut S {
    type Error = S::Error;

    type Assoc2<'a> where 'c: 'a = S::Assoc2<'a>;
}

impl<'c, S : Trait2> Trait2 for &'c mut S {
    type FooFuture<'a, 'b> = impl Future<Output = Result<(), Self::Error>>;
    fn foo<'a: 'b, 'b>(assoc2: &'b mut Self::Assoc2<'a>) -> Self::FooFuture<'a, 'b>
    where
        Self: 'a
    {
        async {
            unimplemented!();
        }
    }
}


fn main() {}
```

oli-obk commented:
No, it is not fixed.

bors commented:
:umbrella: The latest upstream changes (presumably #87141) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk committed:
Use instrument debugging for more readable logs

oli-obk committed:
Remove an unnecessary variable

b-naber commented on compiler/rustc_mir/src/borrow_check/type_check/mod.rs:
```diff
@@ -180,7 +179,66 @@ pub(crate) fn type_check<'mir, 'tcx>(
-            cx.opaque_type_values
+            let mut opaque_type_values = cx.opaque_type_values;
+
+            for (_, revealed_ty) in &mut opaque_type_values {
```
Why does `opaque_type_values` contain elements at the start of the loop? It seems to me that the only place where elements were inserted into this field was in the code you deleted below?

oli-obk replied:
yes, but this function is called multiple times. The opaque types are collected over time.

b-naber commented on compiler/rustc_mir/src/borrow_check/type_check/mod.rs:
```diff
@@ -180,7 +179,66 @@ pub(crate) fn type_check<'mir, 'tcx>(
+                while revealed_ty.has_infer_types_or_consts() {
+                    let prev = *revealed_ty;
+                    trace!(prev=?prev.kind());
+                    let type_resolved = infcx.shallow_resolve(prev);
```
How does this work on types that have only nested inference variables? `shallow_resolve` only resolves inference variables at the top level, so a type that doesn't have an inference variable at the top level would break in this loop.

oli-obk replied:
You're right. I misunderstood what `shallow_resolve` does. I'll implement a visitor as the FIXME above suggests

oli-obk replied:
Turns out there already was a pre-existing visitor that did the right thing™️

b-naber reviewed (commented):
Not sure whether I'm qualified enough to review this, I might be missing something, but there were two parts I didn't understand.

oli-obk committed:
Resolve nested inference variables.

I attempted that with the previous code, but I misunderstdood how
`shallow_resolve` works.

oli-obk committed:
Add a regression test

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3115466046) [(plain)](https://github.com/rust-lang/rust/commit/3d70dee3171606f6d68f2d66b262e652ba916aea/checks/3115466046/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 340bac53c1d3
Successfully tagged rust-ci:latest
Built container sha256:340bac53c1d314aa993546e432d01d64a0e6992532adb5709a9030dee95e547e
Uploading finished image to https://ci-caches.rust-lang.org/docker/b4c226ec1d684c71a4d91fb0cd1a640c97d897266a00433471b3dedeefdb8d321b1df2c8c36c10bd4d6087df0c2b9b507f88d10da95406d6ce945bef9db881e3
upload failed: - to s3://rust-lang-ci-sccache2/docker/b4c226ec1d684c71a4d91fb0cd1a640c97d897266a00433471b3dedeefdb8d321b1df2c8c36c10bd4d6087df0c2b9b507f88d10da95406d6ce945bef9db881e3 Unable to locate credentials
 * branch              master     -> FETCH_HEAD
[CI_JOB_NAME=mingw-check]
[CI_JOB_NAME=mingw-check]
---
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* 340 features
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:13
```

</details>

spastorino commented:
Looks good to me, `bors r=me` once tests are :+1:

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3115639540) [(plain)](https://github.com/rust-lang/rust/commit/5156aab21158fd6a59f729a08fdd4cc7592a88e7/checks/3115639540/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 4e32b0f658c6
Successfully tagged rust-ci:latest
Built container sha256:4e32b0f658c6171bdee6db33a62e06d566f50656706dad01937a269eec19a539
Uploading finished image to https://ci-caches.rust-lang.org/docker/b4c226ec1d684c71a4d91fb0cd1a640c97d897266a00433471b3dedeefdb8d321b1df2c8c36c10bd4d6087df0c2b9b507f88d10da95406d6ce945bef9db881e3
upload failed: - to s3://rust-lang-ci-sccache2/docker/b4c226ec1d684c71a4d91fb0cd1a640c97d897266a00433471b3dedeefdb8d321b1df2c8c36c10bd4d6087df0c2b9b507f88d10da95406d6ce945bef9db881e3 Unable to locate credentials
 * branch              master     -> FETCH_HEAD
[CI_JOB_NAME=mingw-check]
[CI_JOB_NAME=mingw-check]
---
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* 340 features
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:11
```

</details>

oli-obk commented:
@bors r=spastorino

bors commented:
:pushpin: Commit 713044c652a8dc9815df97e1f7a6d167bd22d88c has been approved by `spastorino`

<!-- @bors r=spastorino 713044c652a8dc9815df97e1f7a6d167bd22d88c -->
<!-- homu: {"type":"Approved","sha":"713044c652a8dc9815df97e1f7a6d167bd22d88c","approver":"spastorino"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3115901730) [(plain)](https://github.com/rust-lang/rust/commit/713044c652a8dc9815df97e1f7a6d167bd22d88c/checks/3115901730/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 21da09ca563d
Successfully tagged rust-ci:latest
Built container sha256:21da09ca563dfe8a7cc9364c9c81cf85bfc1c85b0dca537fd2b56f56885660bc
Uploading finished image to https://ci-caches.rust-lang.org/docker/b4c226ec1d684c71a4d91fb0cd1a640c97d897266a00433471b3dedeefdb8d321b1df2c8c36c10bd4d6087df0c2b9b507f88d10da95406d6ce945bef9db881e3
upload failed: - to s3://rust-lang-ci-sccache2/docker/b4c226ec1d684c71a4d91fb0cd1a640c97d897266a00433471b3dedeefdb8d321b1df2c8c36c10bd4d6087df0c2b9b507f88d10da95406d6ce945bef9db881e3 Unable to locate credentials
 * branch                    master     -> FETCH_HEAD
warning: 969c145f96045a156a5776a8f4407e87df5a07b6:.gitmodules, multiple configurations found for 'submodule.src/rust-installer.path'. Skipping second one!
warning: 969c145f96045a156a5776a8f4407e87df5a07b6:.gitmodules, multiple configurations found for 'submodule.src/rust-installer.url'. Skipping second one!
warning: 969c145f96045a156a5776a8f4407e87df5a07b6:.gitmodules, multiple configurations found for 'submodule.src/doc/nomicon.path'. Skipping second one!
---
From https://github.com/rust-lang/miri
 * branch              062cf07d98822beb4c5b1132f47b4397275f403a -> FETCH_HEAD
 * branch              06758c48bd7a77bb5aa43fc50cf344540ba5afef -> FETCH_HEAD
 * branch              dd7f545a69e4b720407e458bf4ade0b207bbf9ee -> FETCH_HEAD
fatal: remote error: upload-pack: not our ref 9ad6e5b8f68ee4bcd85886e50b8b0a70cbb91a52
Errors during submodule fetch:
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

spastorino commented:
@bors retry

spastorino commented:
@bors retry

oli-obk commented:
@bors r=spastorino

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #87136

bors commented:
:pushpin: Commit 713044c652a8dc9815df97e1f7a6d167bd22d88c has been approved by `spastorino`

<!-- @bors r=spastorino 713044c652a8dc9815df97e1f7a6d167bd22d88c -->
<!-- homu: {"type":"Approved","sha":"713044c652a8dc9815df97e1f7a6d167bd22d88c","approver":"spastorino"} -->

spastorino commented:
@bors rollup=never

bors commented:
:hourglass: Testing commit 713044c652a8dc9815df97e1f7a6d167bd22d88c with merge 02263078df37f916ecc500c5a7b6f9227812e22a...
<!-- homu: {"type":"BuildStarted","head_sha":"713044c652a8dc9815df97e1f7a6d167bd22d88c","merge_sha":"02263078df37f916ecc500c5a7b6f9227812e22a"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-nopt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3128807170) [(plain)](https://github.com/rust-lang-ci/rust/commit/02263078df37f916ecc500c5a7b6f9227812e22a/checks/3128807170/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
1 error: lifetime may not live long enough
-   --> $DIR/ret-impl-trait-one.rs:10:80
+   --> $DIR/ret-impl-trait-one.rs:10:65
3    |
- LL |   async fn async_ret_impl_trait1<'a, 'b>(a: &'a u8, b: &'b u8) -> impl Trait<'a> {
-    |  ________________________________--__--__________________________________________^
-    | |                                |   |
-    | |                                |   lifetime `'b` defined here
-    | |                                lifetime `'a` defined here
- LL | |
- LL | |     (a, b)
- LL | | }
-    | |_^ function was supposed to return data with lifetime `'a` but it is returning data with lifetime `'b`
+ LL | async fn async_ret_impl_trait1<'a, 'b>(a: &'a u8, b: &'b u8) -> impl Trait<'a> {
+    |                                --  --                           ^^^^^^^^^^^^^^ opaque type requires that `'b` must outlive `'a`
+    |                                |   |
+    |                                |   lifetime `'b` defined here
+    |                                lifetime `'a` defined here
13    |
14    = help: consider adding the following bound: `'b: 'a`


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/multiple-lifetimes/ret-impl-trait-one.nll/ret-impl-trait-one.nll.stderr
To only update this specific test, also pass `--test-args async-await/multiple-lifetimes/ret-impl-trait-one.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/async-await/multiple-lifetimes/ret-impl-trait-one.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/multiple-lifetimes/ret-impl-trait-one.nll" "-Zborrowck=mir" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2018" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/multiple-lifetimes/ret-impl-trait-one.nll/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: lifetime may not live long enough
  --> /checkout/src/test/ui/async-await/multiple-lifetimes/ret-impl-trait-one.rs:10:65
   |
LL | async fn async_ret_impl_trait1<'a, 'b>(a: &'a u8, b: &'b u8) -> impl Trait<'a> {
   |                                --  --                           ^^^^^^^^^^^^^^ opaque type requires that `'b` must outlive `'a`
   |                                |   |
   |                                |   lifetime `'b` defined here
   |                                lifetime `'a` defined here
   |
   = help: consider adding the following bound: `'b: 'a`
error: aborting due to previous error


------------------------------------------
---
1 error: lifetime may not live long enough
-   --> $DIR/arbitrary_self_types_pin_lifetime_impl_trait-async.rs:8:48
+   --> $DIR/arbitrary_self_types_pin_lifetime_impl_trait-async.rs:8:37
3    |
4 LL |     async fn f(self: Pin<&Self>) -> impl Clone { self }
-    |                          -                     ^^^^^^^^ returning this value requires that `'1` must outlive `'static`
+    |                          -          ^^^^^^^^^^ opaque type requires that `'1` must outlive `'static`
6    |                          |
7    |                          let's call the lifetime of this reference `'1`


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/self/arbitrary_self_types_pin_lifetime_impl_trait-async.nll/arbitrary_self_types_pin_lifetime_impl_trait-async.nll.stderr
To only update this specific test, also pass `--test-args self/arbitrary_self_types_pin_lifetime_impl_trait-async.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/self/arbitrary_self_types_pin_lifetime_impl_trait-async.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/self/arbitrary_self_types_pin_lifetime_impl_trait-async.nll" "-Zborrowck=mir" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2018" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/self/arbitrary_self_types_pin_lifetime_impl_trait-async.nll/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: lifetime may not live long enough
  --> /checkout/src/test/ui/self/arbitrary_self_types_pin_lifetime_impl_trait-async.rs:8:37
   |
LL |     async fn f(self: Pin<&Self>) -> impl Clone { self }
   |                          -          ^^^^^^^^^^ opaque type requires that `'1` must outlive `'static`
   |                          |
   |                          let's call the lifetime of this reference `'1`
   |
help: to allow this `impl Trait` to capture borrowed data with lifetime `'1`, add `'_` as a bound
   |
LL |     async fn f(self: Pin<&Self>) -> impl Clone + '_ { self }

error: aborting due to previous error


---
diff of stderr:

28            found opaque type `impl Sized`
29 
30 error[E0310]: the parameter type `T` may not live long enough
-   --> $DIR/generic_type_does_not_live_long_enough.rs:17:30
32    |
32    |
33 LL | fn wrong_generic<T>(t: T) -> WrongGeneric<T> {


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.full_tait.nll/generic_type_does_not_live_long_enough.full_tait.nll.stderr
To only update this specific test, also pass `--test-args type-alias-impl-trait/generic_type_does_not_live_long_enough.rs`


error in revision `full_tait`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "full_tait" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.full_tait.nll" "-Zborrowck=mir" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.full_tait.nll/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: at least one trait must be specified
  --> /checkout/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.rs:12:24
   |
LL | type WrongGeneric<T> = impl 'static;

warning: the feature `type_alias_impl_trait` is incomplete and may not be safe to use and/or cause compiler crashes
  --> /checkout/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.rs:3:32
   |
   |
LL | #![cfg_attr(full_tait, feature(type_alias_impl_trait))]
   |
   = note: `#[warn(incomplete_features)]` on by default
   = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information


error[E0308]: mismatched types
  --> /checkout/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.rs:9:18
   |
LL |     let z: i32 = x; //~ ERROR mismatched types
   |            ---   ^ expected `i32`, found opaque type
   |            expected due to this
...
...
LL | type WrongGeneric<T> = impl 'static;
   |
   = note:     expected type `i32`
           found opaque type `impl Sized`


error[E0310]: the parameter type `T` may not live long enough
   |
   |
LL | fn wrong_generic<T>(t: T) -> WrongGeneric<T> {
   |
   |
   = help: consider adding an explicit lifetime bound `T: 'static`...

error[E0310]: the parameter type `T` may not live long enough
   |
   |
LL | type WrongGeneric<T> = impl 'static;
   |
   |
   = help: consider adding an explicit lifetime bound `T: 'static`...
   = note: ...so that the type `T` will meet its required lifetime bounds
error: aborting due to 4 previous errors; 1 warning emitted

Some errors have detailed explanations: E0308, E0310.
For more information about an error, try `rustc --explain E0308`.
---
diff of stderr:

19            found opaque type `impl Sized`
20 
21 error[E0310]: the parameter type `T` may not live long enough
-   --> $DIR/generic_type_does_not_live_long_enough.rs:17:30
23    |
23    |
24 LL | fn wrong_generic<T>(t: T) -> WrongGeneric<T> {


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.min_tait.nll/generic_type_does_not_live_long_enough.min_tait.nll.stderr
To only update this specific test, also pass `--test-args type-alias-impl-trait/generic_type_does_not_live_long_enough.rs`


error in revision `min_tait`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "min_tait" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.min_tait.nll" "-Zborrowck=mir" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.min_tait.nll/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: at least one trait must be specified
  --> /checkout/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.rs:12:24
   |
LL | type WrongGeneric<T> = impl 'static;

error[E0308]: mismatched types
  --> /checkout/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.rs:9:18
   |
   |
LL |     let z: i32 = x; //~ ERROR mismatched types
   |            ---   ^ expected `i32`, found opaque type
   |            expected due to this
...
...
LL | type WrongGeneric<T> = impl 'static;
   |
   = note:     expected type `i32`
           found opaque type `impl Sized`


error[E0310]: the parameter type `T` may not live long enough
   |
   |
LL | fn wrong_generic<T>(t: T) -> WrongGeneric<T> {
   |
   |
   = help: consider adding an explicit lifetime bound `T: 'static`...

error[E0310]: the parameter type `T` may not live long enough
   |
   |
LL | type WrongGeneric<T> = impl 'static;
   |
   |
   = help: consider adding an explicit lifetime bound `T: 'static`...
   = note: ...so that the type `T` will meet its required lifetime bounds
error: aborting due to 4 previous errors

Some errors have detailed explanations: E0308, E0310.
For more information about an error, try `rustc --explain E0308`.
---
9    |
10 LL |         |x| x
11    |         ^^^^^

19    = note: expected type `for<'r> Fn<(&'r X,)>`
20               found type `Fn<(&'static X,)>`
21 note: this closure does not fulfill the lifetime requirements
-   --> $DIR/issue-57611-trait-alias.rs:28:9
23    |
24 LL |         |x| x
25    |         ^^^^^



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/issue-57611-trait-alias.min_tait.nll/issue-57611-trait-alias.min_tait.nll.stderr
To only update this specific test, also pass `--test-args type-alias-impl-trait/issue-57611-trait-alias.rs`


error in revision `min_tait`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/type-alias-impl-trait/issue-57611-trait-alias.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "min_tait" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/issue-57611-trait-alias.min_tait.nll" "-Zborrowck=mir" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/issue-57611-trait-alias.min_tait.nll/auxiliary"
------------------------------------------

------------------------------------------
stderr:
---

error[E0308]: mismatched types
  --> /checkout/src/test/ui/type-alias-impl-trait/issue-57611-trait-alias.rs:20:16
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `for<'r> Fn<(&'r X,)>`
              found type `Fn<(&'static X,)>`
note: this closure does not fulfill the lifetime requirements
   |
   |
Some tests failed in compiletest suite=ui compare_mode=Nll mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
LL |         |x| x


error: implementation of `FnOnce` is not general enough
   |
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: closure with signature `fn(&'static X) -> &'static X` must implement `FnOnce<(&'0 X,)>`, for any lifetime `'0`...
   = note: ...but it actually implements `FnOnce<(&'static X,)>`
error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.

---
18    |
19 LL |         |x| x
20    |         ^^^^^

28    = note: expected type `for<'r> Fn<(&'r X,)>`
29               found type `Fn<(&'static X,)>`
30 note: this closure does not fulfill the lifetime requirements
-   --> $DIR/issue-57611-trait-alias.rs:28:9
32    |
33 LL |         |x| x
34    |         ^^^^^



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/issue-57611-trait-alias.full_tait.nll/issue-57611-trait-alias.full_tait.nll.stderr
To only update this specific test, also pass `--test-args type-alias-impl-trait/issue-57611-trait-alias.rs`


error in revision `full_tait`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/type-alias-impl-trait/issue-57611-trait-alias.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "full_tait" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/issue-57611-trait-alias.full_tait.nll" "-Zborrowck=mir" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/issue-57611-trait-alias.full_tait.nll/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
warning: the feature `type_alias_impl_trait` is incomplete and may not be safe to use and/or cause compiler crashes
  --> /checkout/src/test/ui/type-alias-impl-trait/issue-57611-trait-alias.rs:8:32
   |
LL | #![cfg_attr(full_tait, feature(type_alias_impl_trait))]
   |
   = note: `#[warn(incomplete_features)]` on by default
   = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information

---

error[E0308]: mismatched types
  --> /checkout/src/test/ui/type-alias-impl-trait/issue-57611-trait-alias.rs:20:16
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `for<'r> Fn<(&'r X,)>`
              found type `Fn<(&'static X,)>`
note: this closure does not fulfill the lifetime requirements
   |
LL |         |x| x
   |         ^^^^^


error: implementation of `FnOnce` is not general enough
   |
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: closure with signature `fn(&'static X) -> &'static X` must implement `FnOnce<(&'0 X,)>`, for any lifetime `'0`...
   = note: ...but it actually implements `FnOnce<(&'static X,)>`
error: aborting due to 4 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0308`.

---
test result: FAILED. 12020 passed; 6 failed; 125 ignored; 0 measured; 0 filtered out; finished in 87.57s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--host-rustcflags" "-Crpath -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--llvm-version" "12.0.1-rust-1.55.0-nightly" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xray" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always" "--compare-mode" "nll"


Build completed unsuccessfully in 0:16:28
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3128945963)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3128945963","builder_name":"checks-actions"} -->

spastorino commented:
@oli-obk I guess we may need to rebase the PR on top of master and bless tests.

oli-obk committed:
bless nll tests

oli-obk commented:
@bors r=spastorino

bors commented:
:pushpin: Commit d10385242d750288bcbd9e1ab06586217cb2fe2b has been approved by `spastorino`

<!-- @bors r=spastorino d10385242d750288bcbd9e1ab06586217cb2fe2b -->
<!-- homu: {"type":"Approved","sha":"d10385242d750288bcbd9e1ab06586217cb2fe2b","approver":"spastorino"} -->

bors commented:
:hourglass: Testing commit d10385242d750288bcbd9e1ab06586217cb2fe2b with merge b2b7c859c1aae39d26884e760201f5e6c7feeff9...
<!-- homu: {"type":"BuildStarted","head_sha":"d10385242d750288bcbd9e1ab06586217cb2fe2b","merge_sha":"b2b7c859c1aae39d26884e760201f5e6c7feeff9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3141065219)
Approved by: spastorino
Pushing b2b7c859c1aae39d26884e760201f5e6c7feeff9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"spastorino","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3141065219"},"merge_sha":"b2b7c859c1aae39d26884e760201f5e6c7feeff9"} -->
